db_stress_tool/no_batched_ops_stress.cc

```diff
@@ -2376,19 +2379,25 @@ class NonBatchedOpsStressTest : public StressTest {
       Status s;
 
       switch (op) {
-        case Op::Put:
+        case Op::PutOrPutEntity:
         case Op::Merge: {
           ExpectedValue put_value;
           put_value.Put(false /* pending */);
           ryw_expected_values[k] = put_value;
 
+          const uint32_t value_base = put_value.GetValueBase();
+
           char value[100];
-          size_t sz =
-              GenerateValue(put_value.GetValueBase(), value, sizeof(value));
+          const size_t sz = GenerateValue(value_base, value, sizeof(value));
           const Slice v(value, sz);
 
-          if (op == Op::Put) {
-            s = txn->Put(cfh, k, v);
+          if (op == Op::PutOrPutEntity) {
+            if (FLAGS_use_put_entity_one_in > 0 &&
+                (value_base % FLAGS_use_put_entity_one_in) == 0) {
```

Comment:
The value base of a newly created  `ExpectedValue put_value;` should be 0, unlike `PendingExpectedValue pending_expected_value = shared->PreparePut(rand_column_family, rand_key);` which increases the value base by 1 every time we PreparePut().

Then it seems like `(value_base % FLAGS_use_put_entity_one_in) == 0` will always stay true?

Reply:
Thanks for catching this @hx235 ! Will look into this.

P.S. `value_base % FLAGS_use_put_entity_one_in` is `true` if `FLAGS_use_put_entity_one_in==1` but still this is not really what we want :)

Reply:
@hx235 What we're trying to do here is to exercise the read-your-own-write functionality of the transaction layer by adding some queried keys to the transaction, making sure we can read the (uncommitted) values back from the transaction, and then roll the whole thing back. Is there a standard way in the expected state framework for dealing with situations like this? One option I see here is to keep using dedicated `ExpectedValue` objects for the transaction but `SyncPut` some random value to them to make sure we don't always end up with value_base == 1. But maybe there's a better way?

Reply:
Maybe we can do something like this:
```
PendingExpectedValue pending_expected_value = shared->PreparePut(rand_column_family, rand_key);
ExpectedValue expected_value_to_use = pending_expected_value.GetFinalValueBase();
pending_expected_value.Rollback(); // to remove its effect from the expected value system
```
This can get more random expected value by piggy backing PendingExpectedValue but not actually write it to the expected value system.

Reply:
I guess the question is what other threads would see in terms of expected value if we did a `PreparePut`. From a theoretical POV, for any other threads the expected value should be the original value since the transaction hasn't committed (and will never actually commit).

Reply:
> From a theoretical POV, for any other threads the expected value should be the original value

... or rather, the latest value actually written to the db (handwaving here)

Reply:
This digram Hui had for the crash test comes handy https://www.internalfb.com/excalidraw/EX191609

From this diagram: the reading threads should be able to differentiate which value from the expected state to use before and after it's actually written to the DB, the API `PendingExpectedValue::Commit` is used to indicate to the expected state that value has been committed to the DB. In this case, if we use `PendingExpectedValue::Rollback`, it will tell the expected state that value is never actually committed to the DB.  Actually writing the value to the db sounds good too.

Reply:
My feeling is that using `PendingExpectedValue` would work but in a sense it would be too permissive? IIUC what it does is that establishes an acceptable range of values to allow for racing reads/writes (as opposed to the earlier lock-based scheme, where there used to be a single acceptable value). Strictly speaking, the uncommitted value written to the transaction should not be visible to other threads, so it should not affect the acceptable value range, right?

Reply:
I agree, now that I think more about it, the `Prepare` type of operations probably need locking protecting it, that requirement would further change the flow of this test. What we need here is a standalone `ExpectedValue` that has more versatile value options, maybe we can add a constructor and factory method to create a `ExpectedValue` that takes a random value base.

Reply:
Seems like you can already construct an `ExpectedValue` with any value base; or, we can construct an empty one and call `SyncPut` on it (might be nicer)

Reply:
I think that is taking an arbitrary expected value as opposed to value base, which included other parts too. And maybe a random uint64_t is not necessarily a legit expected value. But looks like we can simply construct a default `ExpectedValue`, and use `ExpectedValue::SetValueBase(new_value_base)` to set it to n arbitrary value base.

Reply:
Yeah, seems to me there are many ways to implement this; to me, `SyncPut` seems like the cleanest - it also calls `SetValueBase` internally but it also makes sure that other parts of the bitmask are reset. Not doing so wouldn't cause any issues here I think but still it feels like the right thing to do. (TM) WDYT?

Reply:
I'm little late to the party. I have a question here (mostly for my own learning) @jowlyzhang 

>we can simply construct a default ExpectedValue, and use ExpectedValue::SetValueBase(new_value_base)

Maybe my understanding / reading the code is wrong, but I see that default constructor for ExpectedValue is the following
```
ExpectedValue() : expected_value_(DEL_MASK) {}
```
I'm wondering if this means that if we only call `ExpectedValue::SetValueBase(new_value_base)` without resetting DEL part of it, checking `IsDeleted()` would be true.

+1 to @ltamasi 's approach that resetting other parts of the bitmask by `SyncPut()`